File: jazzer/src/main/java/dev/erst/gridgrind/jazzer/support/JazzerTopology.java
package dev.erst.gridgrind.jazzer.support;

import dev.erst.gridgrind.jazzer.tool.JazzerJson;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/** Loads and validates the committed Jazzer topology shared by runtime support helpers. */
final class JazzerTopology {
  private static final String RESOURCE_PATH =
      "/dev/erst/gridgrind/jazzer/support/jazzer-topology.json";
  private static final Registry REGISTRY = load();

  private JazzerTopology() {}

  static Registry registry() {
    return REGISTRY;
  }

  private static Registry load() {
    TopologyDocument document;
    try {
      document = JazzerJson.readResource(RESOURCE_PATH, TopologyDocument.class);
    } catch (IOException exception) {
      throw new UncheckedIOException("Failed to load Jazzer topology resource", exception);
    }

    Map<String, JazzerHarness> harnessesByKey = new LinkedHashMap<>();
    for (HarnessDocument harnessDocument : document.harnesses()) {
      JazzerHarness harness =
          new JazzerHarness(
              harnessDocument.key(),
              harnessDocument.displayName(),
              harnessDocument.className(),
              harnessDocument.methodName());
      JazzerHarness previous = harnessesByKey.put(harness.key(), harness);
      if (previous != null) {
        throw new IllegalArgumentException("Duplicate Jazzer harness key: " + harness.key());
      }
    }

    Map<String, JazzerRunTarget> runTargetsByKey = new LinkedHashMap<>();
    Map<String, JazzerRunTarget> runTargetsByTaskName = new LinkedHashMap<>();
    for (RunTargetDocument runTargetDocument : document.runTargets()) {
      List<JazzerHarness> harnesses =
          runTargetDocument.harnessKeys().stream()
              .map(
                  key -> {
                    JazzerHarness harness = harnessesByKey.get(key);
                    if (harness == null) {
                      throw new IllegalArgumentException(
                          "Unknown Jazzer harness key in topology: " + key);
                    }
                    return harness;
                  })
              .toList();
      JazzerRunTarget runTarget =
          new JazzerRunTarget(
              runTargetDocument.key(),
              runTargetDocument.displayName(),
              runTargetDocument.taskName(),
              runTargetDocument.workingDirectory(),
              runTargetDocument.activeFuzzing(),
              harnesses);
      JazzerRunTarget previousTarget = runTargetsByKey.put(runTarget.key(), runTarget);
      if (previousTarget != null) {
        throw new IllegalArgumentException("Duplicate Jazzer run target key: " + runTarget.key());
      }
      JazzerRunTarget previousTask = runTargetsByTaskName.put(runTarget.taskName(), runTarget);
      if (previousTask != null) {
        throw new IllegalArgumentException("Duplicate Jazzer task name: " + runTarget.taskName());
      }
      if (runTarget.activeFuzzing()) {
        if (runTarget.harnesses().size() != 1) {
          throw new IllegalArgumentException(
              "Active Jazzer run target must reference exactly one harness: " + runTarget.key());
        }
        if (!runTarget.harnesses().getFirst().key().equals(runTarget.key())) {
          throw new IllegalArgumentException(
              "Active Jazzer run target key must match its harness key: " + runTarget.key());
        }
      }
    }

    Registry registry =
        new Registry(
            List.copyOf(harnessesByKey.values()),
            Map.copyOf(harnessesByKey),
            List.copyOf(runTargetsByKey.values()),
            Map.copyOf(runTargetsByKey),
            Map.copyOf(runTargetsByTaskName));
    return registry;
  }

  record Registry(
      List<JazzerHarness> harnesses,
      Map<String, JazzerHarness> harnessesByKey,
      List<JazzerRunTarget> runTargets,
      Map<String, JazzerRunTarget> runTargetsByKey,
      Map<String, JazzerRunTarget> runTargetsByTaskName) {
    Registry {
      harnesses = List.copyOf(Objects.requireNonNull(harnesses, "harnesses must not be null"));
      harnessesByKey =
          Map.copyOf(Objects.requireNonNull(harnessesByKey, "harnessesByKey must not be null"));
      runTargets = List.copyOf(Objects.requireNonNull(runTargets, "runTargets must not be null"));
      runTargetsByKey =
          Map.copyOf(Objects.requireNonNull(runTargetsByKey, "runTargetsByKey must not be null"));
      runTargetsByTaskName =
          Map.copyOf(
              Objects.requireNonNull(
                  runTargetsByTaskName, "runTargetsByTaskName must not be null"));
    }
  }

  private record TopologyDocument(
      List<HarnessDocument> harnesses, List<RunTargetDocument> runTargets) {
    @SuppressWarnings("UnusedMethod") // Jackson calls this compact constructor reflectively.
    TopologyDocument {
      harnesses = List.copyOf(Objects.requireNonNull(harnesses, "harnesses must not be null"));
      runTargets = List.copyOf(Objects.requireNonNull(runTargets, "runTargets must not be null"));
      if (harnesses.isEmpty()) {
        throw new IllegalArgumentException("Jazzer topology must declare at least one harness");
      }
      if (runTargets.isEmpty()) {
        throw new IllegalArgumentException("Jazzer topology must declare at least one run target");
      }
    }
  }

  private record HarnessDocument(
      String key, String displayName, String className, String methodName) {
    @SuppressWarnings("UnusedMethod") // Jackson calls this compact constructor reflectively.
    HarnessDocument {
      key = requireNonBlank(key, "key");
      displayName = requireNonBlank(displayName, "displayName");
      className = requireNonBlank(className, "className");
      methodName = requireNonBlank(methodName, "methodName");
    }
  }

  private record RunTargetDocument(
      String key,
      String displayName,
      String taskName,
      String workingDirectory,
      boolean activeFuzzing,
      List<String> harnessKeys) {
    @SuppressWarnings("UnusedMethod") // Jackson calls this compact constructor reflectively.
    RunTargetDocument {
      key = requireNonBlank(key, "key");
      displayName = requireNonBlank(displayName, "displayName");
      taskName = requireNonBlank(taskName, "taskName");
      workingDirectory = requireNonBlank(workingDirectory, "workingDirectory");
      harnessKeys =
          List.copyOf(Objects.requireNonNull(harnessKeys, "harnessKeys must not be null"));
      if (harnessKeys.isEmpty()) {
        throw new IllegalArgumentException("harnessKeys must not be empty");
      }
      harnessKeys.forEach(harnessKey -> requireNonBlank(harnessKey, "harnessKey"));
    }
  }

  private static String requireNonBlank(String value, String fieldName) {
    Objects.requireNonNull(value, fieldName + " must not be null");
    String trimmed = value.trim();
    if (trimmed.isEmpty()) {
      throw new IllegalArgumentException(fieldName + " must not be blank");
    }
    return trimmed;
  }
}

File: gradle/build-logic/src/main/kotlin/dev/erst/gridgrind/buildlogic/JazzerTopology.kt
package dev.erst.gridgrind.buildlogic

import groovy.json.JsonSlurper
import org.gradle.api.Project

internal data class JazzerHarnessSpec(
    val key: String,
    val displayName: String,
    val className: String,
    val methodName: String,
) {
    fun inputResourceDirectory(): String {
        val packagePath = className.substringBeforeLast('.').replace('.', '/')
        val simpleName = className.substringAfterLast('.')
        return "$packagePath/${simpleName}Inputs/$methodName"
    }
}

internal data class JazzerRunTargetSpec(
    val key: String,
    val displayName: String,
    val taskName: String,
    val workingDirectory: String,
    val activeFuzzing: Boolean,
    val harnessKeys: List<String>,
)

internal data class JazzerTopology(
    val harnesses: List<JazzerHarnessSpec>,
    val runTargets: List<JazzerRunTargetSpec>,
) {
    private val harnessesByKey: Map<String, JazzerHarnessSpec> = harnesses.associateBy(JazzerHarnessSpec::key)
    private val runTargetsByKey: Map<String, JazzerRunTargetSpec> = runTargets.associateBy(JazzerRunTargetSpec::key)

    fun harness(key: String): JazzerHarnessSpec =
        harnessesByKey[key] ?: throw IllegalArgumentException("Unknown Jazzer harness key: $key")

    fun runTarget(key: String): JazzerRunTargetSpec =
        runTargetsByKey[key] ?: throw IllegalArgumentException("Unknown Jazzer run target key: $key")

    companion object {
        private const val TOPOLOGY_PATH = "src/main/resources/dev/erst/gridgrind/jazzer/support/jazzer-topology.json"

        fun load(project: Project): JazzerTopology {
            val topologyFile = project.layout.projectDirectory.file(TOPOLOGY_PATH).asFile
            val root = JsonSlurper().parse(topologyFile) as? Map<*, *>
                ?: throw IllegalArgumentException("Jazzer topology must be a JSON object: $TOPOLOGY_PATH")
            val harnesses =
                maps(root["harnesses"], "harnesses").map { map ->
                    JazzerHarnessSpec(
                        key = string(map, "key"),
                        displayName = string(map, "displayName"),
                        className = string(map, "className"),
                        methodName = string(map, "methodName"),
                    )
                }
            val runTargets =
                maps(root["runTargets"], "runTargets").map { map ->
                    JazzerRunTargetSpec(
                        key = string(map, "key"),
                        displayName = string(map, "displayName"),
                        taskName = string(map, "taskName"),
                        workingDirectory = string(map, "workingDirectory"),
                        activeFuzzing = boolean(map, "activeFuzzing"),
                        harnessKeys = strings(map["harnessKeys"], "harnessKeys"),
                    )
                }
            require(harnesses.isNotEmpty()) { "Jazzer topology must declare at least one harness" }
            require(harnesses.map(JazzerHarnessSpec::key).toSet().size == harnesses.size) {
                "Jazzer topology harness keys must be unique"
            }
            require(runTargets.map(JazzerRunTargetSpec::key).toSet().size == runTargets.size) {
                "Jazzer topology run target keys must be unique"
            }
            val topology = JazzerTopology(harnesses = harnesses, runTargets = runTargets)
            topology.validate()
            return topology
        }

        private fun string(map: Map<*, *>, key: String): String =
            map[key] as? String ?: throw IllegalArgumentException("Missing string field '$key' in Jazzer topology")

        private fun boolean(map: Map<*, *>, key: String): Boolean =
            map[key] as? Boolean ?: throw IllegalArgumentException("Missing boolean field '$key' in Jazzer topology")

        private fun maps(value: Any?, key: String): List<Map<*, *>> =
            (value as? List<*>)?.map { item ->
                item as? Map<*, *> ?: throw IllegalArgumentException("Field '$key' must contain JSON objects")
            } ?: throw IllegalArgumentException("Missing array field '$key' in Jazzer topology")

        private fun strings(value: Any?, key: String): List<String> =
            (value as? List<*>)?.map { item ->
                item as? String ?: throw IllegalArgumentException("Field '$key' must contain strings")
            } ?: throw IllegalArgumentException("Missing array field '$key' in Jazzer topology")
    }

    private fun validate() {
        runTargets.forEach { target ->
            require(target.harnessKeys.isNotEmpty()) {
                "Jazzer run target '${target.key}' must reference at least one harness"
            }
            target.harnessKeys.forEach { harness(it) }
            if (target.activeFuzzing) {
                require(target.harnessKeys.size == 1) {
                    "Active Jazzer run target '${target.key}' must reference exactly one harness"
                }
                require(target.harnessKeys.single() == target.key) {
                    "Active Jazzer run target '${target.key}' must share its key with its harness"
                }
            }
        }
    }
}
